RavenDB-21027 Corax performance optimizations

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21027
https://issues.hibernatingrhinos.com/issue/RavenDB-21059
https://issues.hibernatingrhinos.com/issue/RavenDB-21004
https://issues.hibernatingrhinos.com/issue/RavenDB-21065

### Additional description

Various performance optimizations discovered during large scale benchmarks.
See individual commits for the details

redknightlois committed:
RavenDB-21004: Removed multiple calls to AssertContextNotDisposed

Most of the calls to `AssertContextNotDisposed` where actually useless as protected and private methods would have already called in from the public interface.

ayende committed:
RavenDB-21027 - Avoid zeroing memory on the container on delete (saves 2% during indexing)

redknightlois committed:
RavenDB-21004: Improved code generation

redknightlois committed:
RavenDB-21004: Ensure stackalloc buffers that don't need zeroing aren't zeroed.

redknightlois committed:
RavenDB-21004: Avoid struct copies.

ayende committed:
RavenDB-21027 Only raise the NeedToSortToDeleteDuplicates flag if we detect that we really need to

ayende committed:
RavenDB-21027 Prepare lists for higher perf

ayende committed:
RavenDB-21027 Change the way we are doing merges for posting list.
Previously, we were doing a three way merge between Existing, Additions & Removals.
Now we are marking removals with the first bit set (taking advantage of the fact that posting list values *must* have their first two bits cleared).
Merging those lists is then done by appending, sorting and removing of duplicates, taking care to remove those with the first bit set

ayende committed:
RavenDB-21027 Fixing tests

ayende committed:
RavenDB-21027 Apply new index skipping existing document updates to Corax as well as Lucene.

redknightlois commented on src/Corax/IndexWriter.cs:
```diff
@@ -2070,8 +2076,15 @@ private AddEntriesToTermResult AddEntriesToTermResultViaLargePostingList(ref Ent
-            entries.GetEncodedAdditionsAndRemovals(_entriesAllocator, out Span<long> additions, out Span<long> removals);
-            var numberOfEntries = PostingList.Update(_transaction.LowLevelTransaction, ref postingListState, additions, removals);
+            entries.GetEncodedAdditionsAndRemovals(_entriesAllocator, out var additions, out var removals);
+            //TEMP CODE
```
Not temp anymore :)

ayende replied:
Thanks

redknightlois commented on src/Voron/Data/PostingLists/PostingListLeafPage.cs:
```diff
@@ -81,29 +82,23 @@ public static void InitLeaf(PostingListLeafPageHeader* header)
+        var needSorting = removalsCount > 0 || // any removal force sorting
+                          // here we test if the first new addition is smaller than the largest existing, requiring sorting  
+                          (maxAdditionsLimit > 0 && additions[0] <= tempList.RawItems[tempList.Count - 1]);
 
```
This section needs very important comment we want to avoid the "When I wrote this code only god and I knew what it did. Now only god knows".

ayende replied:
Done

maciejaszyk commented:
LGTM

redknightlois committed:
RavenDB-21027: Optimizing space creation on pages.

redknightlois committed:
RavenDB-21027: Micro optimization of adding to pages.

ayende committed:
RavenDB-21059 PostingList handling of large lists result in deep trees

ayende committed:
RavenDB-21027 Adding comments explaining how we do list merging between existing, additions & removals

ayende committed:
RavenDB-21027 Optimizing how we are updating entries for term, we now buffer them in memory, sort them and then add in a bulk manner.

arekpalinski commented:
All of the test builds have crashed on this PR:

https://jenkins.hibernatingrhinos.com/job/PR_Tests/7676/consoleFull
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7675/consoleFull
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7674/consoleFull
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7666/consoleFull
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7665/consoleFull

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
at Corax.IndexWriter.InsertEntriesForTerm(Int64)
at Corax.IndexWriter.InsertTextualField(Voron.Data.BTrees.Tree, IndexedField, System.Span`1<Byte>, Voron.Slice[] ByRef)
at Corax.IndexWriter.Commit[[Corax.Utils.EmptyStatsScope, Corax, Version=6.0.0.0, Culture=neutral, PublicKeyToken=null]](Corax.Utils.EmptyStatsScope)
at Corax.IndexWriter.Commit()
at Corax.IndexWriter.<TryDeleteEntry>g__FlushBatch|75_0(<>c__DisplayClass75_0 ByRef)
at Corax.IndexWriter.TryDeleteEntry(Voron.Slice, Int64 ByRef)
at Corax.IndexWriter.Update(System.ReadOnlySpan`1<Byte>)
at Raven.Server.Documents.Indexes.Persistence.Corax.CoraxIndexWriteOperation.UpdateDocument(Sparrow.Json.LazyStringValue, Sparrow.Json.LazyStringValue, System.Object, Raven.Server.Documents.Indexes.IndexingStatsScope, Sparrow.Json.JsonOperationContext)
at Raven.Server.Documents.Indexes.MapReduce.ReduceMapResultsBase`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].OutputReduceResultsCorax(Raven.Server.ServerWide.Context.TransactionOperationContext, Raven.Server.Documents.Indexes.IndexingStatsScope, System.Lazy`1<Raven.Server.Documents.Indexes.Persistence.IndexWriteOperationBase>, Sparrow.Json.LazyStringValue, System.Collections.Generic.IEnumerable`1<System.Object>, Boolean)
at Raven.Server.Documents.Indexes.MapReduce.ReduceMapResultsBase`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].HandleTreeReduction(Raven.Server.ServerWide.Context.TransactionOperationContext, Raven.Server.Documents.Indexes.IndexingStatsScope, Raven.Server.Documents.Indexes.MapReduce.MapReduceResultsStore, Voron.Impl.LowLevelTransaction, System.Lazy`1<Raven.Server.Documents.Indexes.Persistence.IndexWriteOperationBase>, Sparrow.Json.LazyStringValue, Voron.Data.Tables.Table, System.Threading.CancellationToken)
at Raven.Server.Documents.Indexes.MapReduce.ReduceMapResultsBase`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].HandleTreeReduction(Raven.Server.ServerWide.Context.TransactionOperationContext, Raven.Server.Documents.Indexes.IndexingStatsScope, Raven.Server.Documents.Indexes.MapReduce.MapReduceResultsStore, Voron.Impl.LowLevelTransaction, System.Lazy`1<Raven.Server.Documents.Indexes.Persistence.IndexWriteOperationBase>, Sparrow.Json.LazyStringValue, Voron.Data.Tables.Table, System.Threading.CancellationToken)
at Raven.Server.Documents.Indexes.MapReduce.ReduceMapResultsBase`1[[System.__Canon, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Execute(Raven.Server.Documents.Indexes.QueryOperationContext, Raven.Server.ServerWide.Context.TransactionOperationContext, System.Lazy`1<Raven.Server.Documents.Indexes.Persistence.IndexWriteOperationBase>, Raven.Server.Documents.Indexes.IndexingStatsScope, System.Threading.CancellationToken)
at Raven.Server.Documents.Indexes.Index.DoIndexingWork(Raven.Server.Documents.Indexes.IndexingStatsScope, System.Threading.CancellationToken)
at Raven.Server.Documents.Indexes.Index.ExecuteIndexing()
at Raven.Server.Documents.Indexes.Index.<StartIndexingThread>b__150_0(System.Object)
at Raven.Server.Utils.PoolOfThreads+PooledThread.DoWork()
at Raven.Server.Utils.PoolOfThreads+PooledThread.Run()
at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
```

ayende commented:
Looks like an error when running `Should_correctly_reduce_after_updating_all_documents`

ayende committed:
RavenDB-21027 Fixing renewal of native lists after we reset the writer

ayende commented:
Found the issue, will have a fix shortly

arekpalinski commented:
Now it managed to run all tests. You can see the failures here:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7693/testReport/

maciejaszyk committed:
RavenDB-21065 Extend TermProvider API to allow reading terms in big batches.

maciejaszyk committed:
RavenDB-21065 Push MultiTermMatches (when it has more terms than TermMatch has documents)  to upper-left node in BinaryMatch to avoid calling Fill/Reset multiple times for unique values.

maciejaszyk committed:
RavenDB-21065 Save copy of memoization match to avoid losing query state.

maciejaszyk committed:
RavenDB-21065 Decrease depth of binary match in `In` primitive.

maciejaszyk committed:
RavenDB-21065 MultiTermMatch - allow to fetch terms in batches without using TermMatches.

ayende committed:
RavenDB-21027 Fixing tests that passed wrong numbers to the posting lists

ayende commented:
Great, those are minor test fixes, nothing major

ayende committed:
RavenDB-21065 Ensure that we are clearing the cursor before pushing to it.

ayende committed:
RavenDB-21065 Fixing an error where we would be setting the _current value to Start *after* we detected that there are no values there

ayende committed:
RavenDB-21065
- Fixing missing _isFinished in BackwardIterator leading to additional scanning of results
- Implemented EqualsMatchComparer for MultiUnaryMatch
- Verify we are always working with write tx.

ayende committed:
RavenDB-21065 Detecting whatever there is a previous element when we scan, so we know if the next value we got has exceeded the maximum we need

maciejaszyk committed:
RavenDB-21027 Reset LastSearchPosition when performing Reset on iterator.

ayende committed:
RavenDB-21027 When we run into a large AndWith, we'll memoize the value and instead of calling AndWith multiple times

ayende committed:
RavenDB-21027 Properly call memoization on AndWith query

ayende committed:
RavenDB-21027 Reduce the number of allocations during query by reusing the buffer

ayende committed:
RavenDB-21027 Reducing allocation during scanning of large sets

ayende committed:
RavenDB-21027 Fixing how we detect that we need to re-create the buffer

maciejaszyk committed:
RavenDB-21065 UnaryMatches:
- multiple terms in field
- no fields at all

maciejaszyk committed:
RavenDB-21065 AllIn: query building fix.

maciejaszyk committed:
RavenDB-21065 Exists: batch fills

maciejaszyk committed:
RavenDB-21027 UnaryMatch `AllIn` implementation requires sorted terms.

maciejaszyk committed:
RavenDB-21027 Clear the stack-allocated array before passing it to the bitset, as it may contain high bits from previous stack data.

maciejaszyk committed:
RavenDB-21027 Disable warning with pragma
